lib/graphql/schema/build_from_ast.rb

```diff
@@ -0,0 +1,260 @@
+module GraphQL
+  class Schema
+    # You can use the result of {GraphQL.parse} to make a schema.
+    # This schema is missing some important details like
+    # `resolve` functions, but it does include the full type system,
+    # so you can use it to validate queries.
+    module BuildFromAST
+      extend self
+
+      class InvalidDocument < Error; end;
+
+      # Create schema with the AST of a parsed schema.
+      # @param document [GraphQL::Language::Nodes::Document] A Document node obtained from {GraphQL.parse}
+      # @return [GraphQL::Schema] the schema described by `document`
+      def build(document)
+        raise InvalidDocument.new('Must provide a document ast.') if !document || !document.is_a?(GraphQL::Language::Nodes::Document)
+
+        schema_definition = nil
+        types = {}
+        directives = {}
+        type_resolver = -> (type) { -> { resolve_type(types, type) } }
```

Comment:
Copied this logic from `Schema::Loader`.

The alternative would be to iterate over `document.definitions` twice. Once to initialize `types` with empty definitions that way when you iterate a second time you can reference objects instead of passing a proc that resolves later.